src/main/utility/sockaddr.rs

```diff
@@ -0,0 +1,688 @@
+    /// filesystem path.
+    pub fn as_path(&self) -> Option<&CStr> {
+        let path = self.sun_path()?;
+
+        if path.is_empty() || path[0] == 0 {
+            return None;
+        }
+
+        let first_nul = path.iter().position(|&x| x == 0)?;
+
+        return Some(CStr::from_bytes_with_nul(&path[..(first_nul + 1)]).unwrap());
+    }
+
+    /// If the socket address represents an abstract address, returns the bytes representing the
+    /// name of the abstract socket address. These bytes do not include the nul byte at
+    /// `sun_path[0]`.
+    pub fn as_abstract(&self) -> Option<&[u8]> {
+        let name = self.sun_path()?;
+
+        if name.is_empty() {
+            return None;
+        }
+
+        if name[0] != 0 {
+            return None;
+        }
+
+        return Some(&name[1..]);
+    }
+
+    /// Is the unix socket address unnamed? On Linux, unnamed unix sockets are unnamed if their
+    /// length is `size_of::<libc::sa_family_t>()`.
+    pub fn is_unnamed(&self) -> bool {
+        (self.len as usize) == memoffset::span_of!(libc::sockaddr_un, sun_family).end
+    }
+
+    /// Returns a slice with the valid bytes of `sun_path`, or `None` if the address length is too
+    /// short.
+    fn sun_path(&self) -> Option<&[u8]> {
+        let path_offset = memoffset::offset_of!(libc::sockaddr_un, sun_path);
+        let path_len = (self.len as usize).checked_sub(path_offset)?;
+
+        Some(i8_to_u8_slice(&self.addr.borrow().sun_path[..path_len]))
+    }
+
+    /// Get an owned unix socket address.
+    pub fn into_owned(self) -> SockaddrUnix<libc::sockaddr_un> {
+        SockaddrUnix {
+            addr: *self.addr.borrow(),
+            len: self.len,
+        }
+    }
+
+    /// Get a borrowed unix socket address.
+    pub fn as_ref(&self) -> SockaddrUnix<&libc::sockaddr_un> {
+        SockaddrUnix {
+            addr: self.addr.borrow(),
+            len: self.len,
+        }
+    }
+
+    /// Get a pointer to the unix socket address. All fields of the `libc::sockaddr_un` will be
+    /// properly initialized.
+    pub fn as_ptr(&self) -> (*const libc::sockaddr_un, libc::socklen_t) {
+        (self.addr.borrow(), self.len)
+    }
+}
+
+impl<T> SockaddrUnix<T>
+where
+    T: BorrowMut<libc::sockaddr_un>,
+{
+    /// Get a mutably borrowed unix socket address.
+    pub fn as_mut(&mut self) -> SockaddrUnix<&mut libc::sockaddr_un> {
+        SockaddrUnix {
+            addr: self.addr.borrow_mut(),
+            len: self.len,
+        }
+    }
+}
+
+impl SockaddrUnix<libc::sockaddr_un> {
+    /// Get a new `SockaddrUnix` with the given path. Will return `None` if the path is empty or is
+    /// too large.
+    pub fn new_path(path: &CStr) -> Option<Self> {
+        let path = path.to_bytes();
+
+        // this should be guaranteed by the CStr
+        debug_assert!(!path.contains(&0));
+
+        if path.is_empty() {
+            // you cannot have a pathname unix socket address with no path
```

Comment:
When a conversion might fail in multiple ways like this, it might be useful for debugging to return a Result with a custom error type... or maybe just log at `trace` or `debug` level

Reply:
I agree custom error types would be nice, but didn't want to add them yet until we were happy with a design. I can add them in a new PR though. I decided against adding log messages since we might use code like:

```rust
if let Some(path) = addr.as_path() {
  host.filesystem.bind(path, sock);
} else if let Some(name) = addr.as_abstract() {
  host.abstract_namespace.bind(name, sock);
} else {
  ...
]
```

And if we log on every failed `as_` method call, we might end up with a lot of unneeded log messages.